### ISSUE TITLE
Can combine skip intermediate calculations for sync values?

### ISSUE THREAD

**Issue by standy** [cujojs/most#227]
When streams are syncronized and im trying to combine them, there is intermediate calculations, which im trying to avoid

Here is example: http://requirebin.com/?gist=ed0ccb7df29f651b13a63fa83a59ecdc
```
var a$ = most.periodic(1000).delay(1000).scan(x => x + 1, 0).take(5);
var b$ = a$.map(x => x + 1);
most.combine(sum, a$, b$)
	//.debounce(0)
	.forEach(x => console.log('a+b=', x, new Date().toISOString()));

function sum(a, b) { return a + b; } 
```
And a diagram:
```
a$            -0--1--2--3--4--
b$            -1--2--3--4--5--
sum:expected  -1--3--5--7--9--
sum:real      -1--23-45-67-89-
```
On 1st second i have a=1 and b=2, so i expect sum=3, but i had sum=2 and then sum=3

Using `debounce` will help, but its a bad decision to my mind. Becouse it slow, unrealiable, and may lead to race conditions (for example, if next i want to combine sum and a$), etc

What if combine, or a new method, can detect when variables are calculating and push only when last of them done?

**Comment by axefrog** [comment]
If you know there's a 1:1 correlation between the data in each stream, using `zip`, which is the same as combine, but with synchronised output. That is the function you pass it will be called once for each complete set of emissions on each channel.

**Comment by standy** [comment]
Yes, but only for 1:1, so zip will fail when you try to filter one of streams
```
a$            -0--1--2--3--4--
b$            -1-----3-----5--
sum:expected  -1-----5-----9--
zip           -1-----4-----7--
```

**Comment by axefrog** [comment]
I guess it's just a case of the fact that you can't have it both ways. How would most know what you consider to be an intermediate calculation? Would `sample` solve the problem?

**Comment by briancavalier** [comment]
Hi @standy.  Thanks for jumping in with good info, @axefrog!

As you noticed, in push-based implementations, simultaneity poses an interesting problem.  It's much less of a problem in pull-based impls (e.g. [catalyst](https://github.com/briancavalier/catalyst) handles it in its [mergeWith operator](https://github.com/briancavalier/catalyst/blob/master/src/event.js#L70-L71))

> can detect when variables are calculating and push only when last of them done?

That is the essence of the problem: *How* would one detect when variables are calculating?

I do believe that there is a reasonable solution to be had ... we just haven't managed to figure it out yet :)

One way might be to wait "long enough" to deem that all events in the "current time unit" (whatever that unit is, millisecond, etc.) must have occurred.  But how long is long enough?  If the time unit is millisecond, then we'd need at least need to wait until the end of the current millisecond.  While most's scheduler *does* have millisecond resolution, this still poses a few interesting problems:

1. It means that event propagation through combine() *cannot* be synchronous anymore.  combine must use the scheduler, so will incur a small, but non-zero perf penalty.

2. Things get really interesting when you stack multiple combines: `combine(f1, s1, combine(f2, s2, combine(f3, s3, combine(f4, s4, s5))))`  If the innermost combine waits until the end of the current millisecond, how long should each of the outer combines wait?  Is the perf penalty cumulative?

I'd love to find some key insights that allow a solution to be implemented efficiently, so if you have thoughts/suggestions/etc. I'd certainly love to hear them.

**Comment by briancavalier** [comment]
@standy In your specific example, zip would work, as @axefrog mentioned.  But like you said, if you change the proportion of one of the streams, i.e. via filter, throttle, etc. zip wouldn't work.  One possible solution would be to zip first, and then filter afterward.

**Comment by standy** [comment]
Did some research, and now im pretty sure it is possible with this implementation. 
Actually, when there is a list of values that about to start calculating, you could mark them all as "not done yet", and unmark one by one right after. So inside combine method all you need is don't push if there is more then one marked value.
My commit is just dirty hack, but i hope it helps

**Comment by briancavalier** [comment]
@standy Thanks for investigating.  Unfortunately, that approach won't work in all cases.  The scheduler is a *task* scheduler, and not an *event* scheduler.  Not all events pass through the scheduler, and sometimes many events are produced by a single task.  As one example, DOM events (in fact, any `EventTarget`) [don't use the scheduler at all](https://github.com/cujojs/most/blob/master/lib/source/EventTargetSource.js#L16-L25), since they are naturally async and most.js doesn't need to introduce extra asynchrony.  So, something like this wouldn't detect simultaneity:

```js
const s = fromEvent('input', textBox)
combine(add2, s, s.map(x => x+1))
```

I read a paper recently (can't find the link atm, but I'll post it here if I find it) that proposed a solution.  The author's observation was that in a single-threaded push-based system, simultaneous events can only occur when they originated from the same source.  The proposed solution was to tag every originating event source (i.e. fromEvent) with an id.  They proposed to use that to determine if events from multiple inputs had ultimately come from the same source, and thus might be simultaneous.

I'm not convinced it can work in all cases in the face of filtering (and other combinators that drop or insert new events), but still an interesting idea.
